### PR TITLE
[Train] Implement `AccelerateTrainer`

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Implements the `AccelerateTrainer`, providing an integration with Hugging Face Accelerate in Ray Train.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/requirements/ml/requirements_train.txt [hunk 1]
```diff
@@ -12,6 +12,7 @@ tensorboardX==2.4.1
 # `python/ray/air/tests/test_huggingface_trainer.py`
 transformers==4.18.0; python_version <= '3.6'
 transformers==4.19.1; python_version > '3.6'
-accelerate==0.5.1
+accelerate==0.5.1; python_version <= '3.6'
+accelerate==0.17.1; python_version > '3.6'
```
Comment (richardliaw) [inline python/requirements/ml/requirements_train.txt:16]: Can you make sure we document that this is what we test against? You can just list the 0.17.1 version.

Comment (Yard1) [reply python/requirements/ml/requirements_train.txt:16]: where should we document that? here, in the test file, or in the docs?

Comment (richardliaw) [reply python/requirements/ml/requirements_train.txt:16]: docstring ?

**Review by richardliaw — APPROVED** [review]

File: doc/source/ray-air/trainers.rst [hunk 2]
```diff
@@ -140,8 +140,11 @@ To use this trainer, you will need to first run ``pip install -U lightgbm-ray``.
 Other Trainers
 --------------
 
-HuggingFace Trainer
-~~~~~~~~~~~~~~~~~~~
+HuggingFace
+~~~~~~~~~~~
```
Comment (matthewdeng) [inline doc/source/ray-air/trainers.rst:144]: nit since you changed the others?
```suggestion
Hugging Face
~~~~~~~~~~~~
```

File: doc/source/ray-air/trainers.rst [hunk 3]
```diff
@@ -164,6 +167,29 @@ training via Pytorch DDP.
...
+
+If you prefer a more fine-grained Hugging Face API than what Transformers provides, you can use :class:`AccelerateTrainer <ray.train.huggingface.accelerate.AccelerateTrainer>`
+to run training functions making use of Hugging Face Accelerate. This Trainer is also an extension of  :class:`TorchTrainer <ray.train.torch.TorchTrainer>`, allowing
+you to pass an Accelerate configuration file generated with ``accelerate config`` to be applied on all training workers. This ensures that the worker
+environments are set up correctly for Accelerate, allowing you to take advantage of Accelerate APIs and integrations such as DeepSpeed and FSDP
+just as you would if you were running Accelerate without Ray.
+
+.. note::
+    ``AccelerateTrainer`` will override some settings set with ``accelerate config``, mainly related to
+    the topology and networking. See the ``AccelerateTrainer`` API reference for more details.
```
Comment (matthewdeng) [inline doc/source/ray-air/trainers.rst:181]: Link to the reference directly here?

File: python/ray/train/huggingface/accelerate/accelerate_trainer.py [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+            self._accelerate_config_raw = self.accelerate_config.config_raw
+            self._deepspeed_config_file_raw = (
+                self.accelerate_config.deepspeed_config_raw
+            )
+            train_loop_config = self._create_train_loop_per_worker(
+                train_loop_per_worker,
+                train_loop_config,
+                self._accelerate_config_raw,
+                self._deepspeed_config_file_raw,
+            )
```
Comment (matthewdeng) [inline python/ray/train/huggingface/accelerate/accelerate_trainer.py:306]: Move this outside of the if/else?

File: python/ray/train/huggingface/accelerate/accelerate_trainer.py [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+        )
+        for key in train_loop_per_worker_config:
+            if key in RESERVED_KEYS:
+                raise ValueError(
+                    f"'{key}' is a reserved key in `train_loop_per_worker_config`."
+                )
+        if train_loop_per_worker:
+            train_loop_per_worker_config[
+                TRAIN_LOOP_PER_WORKER_KEY
+            ] = train_loop_per_worker
```
Comment (matthewdeng) [inline python/ray/train/huggingface/accelerate/accelerate_trainer.py:358]: Is there a cleaner way such that we don't have to do this... 

Preferably we can create the final `train_loop_per_worker` from the input `train_loop_per_worker` and accelerate configs, without using `config` as a container.

Comment (Yard1) [reply python/ray/train/huggingface/accelerate/accelerate_trainer.py:358]: sure - that was actually the approach I took first!

File: python/ray/train/huggingface/accelerate/accelerate_trainer.py [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+    ):
+
+        if Version(accelerate.__version__) < Version("0.17.0.dev0"):
+            raise RuntimeError(
+                "AccelerateTrainer requires accelerate>=0.17.0, "
+                f"got {accelerate.__version__}"
+            )
+
+        self.accelerate_config = accelerate_config
+        if isinstance(self.accelerate_config, AccelerateConfigWrapper):
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/train/huggingface/accelerate/accelerate_trainer.py:296]: It is internal, that's correct. The logic is:
1. On first init (which happens in the process the user started), we read the file and wrap the contents in the internal `AccelerateConfigWrapper`
2. The Trainer class is initialized again inside the Trainable. We read the `accelerate_config` argument and see that it is wrapped, in which case we do not try to load the file from disk again and instead use the string as raw contents.

In other words, this is necessary to deal with the issue of the Trainer being initialized twice (once by the user, the second time in the trainable).

File: python/ray/train/huggingface/accelerate/accelerate_trainer.py [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+
+    def training_loop(self) -> None:
+        old_train_loop_per_worker = self._train_loop_per_worker
+        self._train_loop_per_worker = self._wrap_train_loop_per_worker(
+            self._train_loop_per_worker,
+            self._accelerate_config_raw,
+            self._deepspeed_config_file_raw,
+        )
+        try:
+            ret = super().training_loop()
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/train/huggingface/accelerate/accelerate_trainer.py:340]: We wrap the user-defined training loop function in our function to apply the Accelerate config and prepare the environment.

File: python/ray/train/huggingface/accelerate/accelerate_trainer.py [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+        deepspeed_config_file_raw: str,
+    ):
+        """Wrap around train_loop_per_worker to set necessary Accelerate env vars."""
+
+        @functools.wraps(train_loop_per_worker)
+        def _accelerate_train_loop_per_worker(*args, **kwargs):
+            with tempfile.TemporaryDirectory() as tempdir:
+                # Write Accelerate config to file so it can be read
+                # by Accelerate
+                temp_config_file = os.path.join(tempdir, "default_config.yaml")
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/train/huggingface/accelerate/accelerate_trainer.py:359]: I'd rather keep it as close to accelerate as possible and not introduce extra abstractions on our side

File: doc/requirements-doc.txt [hunk 9]
```diff
@@ -1,6 +1,7 @@
 # Production requirements. This is what readthedocs.org picks up
 
 # Python / ML libraries
+accelerate>=0.17.0
```
Comment (gjoliver) [inline doc/requirements-doc.txt:4]: why do we need the actual accelerate library for the docs?

File: doc/source/ray-air/doc_code/accelerate_trainer.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+
+# If using GPUs, set this to True.
+use_gpu = False
+
+
+input_size = 1
+layer_size = 15
+output_size = 1
+num_epochs = 3
```
Comment (gjoliver) [inline doc/source/ray-air/doc_code/accelerate_trainer.py:19]: turn these into flags?
alternatively, they should be ALL_CAP if these are constants

File: doc/source/ray-air/doc_code/accelerate_trainer.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+            inputs, labels = torch.unsqueeze(batches["x"], 1), batches["y"]
+            output = model(inputs)
+            loss = loss_fn(output, labels)
+            optimizer.zero_grad()
+            accelerator.backward(loss)
+            optimizer.step()
+            print(f"epoch: {epoch}, loss: {loss.item()}")
+
+        session.report(
+            {},
```
Comment (gjoliver) [inline doc/source/ray-air/doc_code/accelerate_trainer.py:55]: can you add comments so we know what this empty dict is?
it's metrics dict right? maybe it's better to send loss as a metric

File: doc/source/ray-air/trainers.rst [hunk 12]
```diff
@@ -164,6 +167,30 @@ training via Pytorch DDP.
         :start-after: __hf_trainer_start__
         :end-before: __hf_trainer_end__
 
+Accelerate
+**********
+
+If you prefer a more fine-grained Hugging Face API than what Transformers provides, you can use :class:`AccelerateTrainer <ray.train.huggingface.accelerate.AccelerateTrainer>`
+to run training functions making use of Hugging Face Accelerate. This Trainer is also an extension of  :class:`TorchTrainer <ray.train.torch.TorchTrainer>`, allowing
+you to pass an Accelerate configuration file generated with ``accelerate config`` to be applied on all training workers. This ensures that the worker
```
Comment (gjoliver) [inline doc/source/ray-air/trainers.rst:175]: curious, does this functionality (i.e. being able to pass a config file to all training workers) actually has anything to do with TorchTrainer?

File: python/ray/train/BUILD [hunk 13]
```diff
@@ -586,7 +594,7 @@ py_test(
 
 py_test(
     name = "test_trainer_restore",
-    size = "medium",
+    size = "large",
```
Comment (gjoliver) [inline python/ray/train/BUILD:597]: why change this? does this PR have any impact on this test?

File: python/ray/train/huggingface/accelerate/_accelerate_utils.py [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+        args.deepspeed_fields_from_accelerate_config = ",".join(
+            args.deepspeed_fields_from_accelerate_config
+        )
+        deepspeed_launcher(args)
+    elif args.use_fsdp and not args.cpu:
+        multi_gpu_launcher(args)
+    elif args.use_megatron_lm and not args.cpu:
+        multi_gpu_launcher(args)
+    elif args.multi_gpu and not args.cpu:
+        multi_gpu_launcher(args)
```
Comment (gjoliver) [inline python/ray/train/huggingface/accelerate/_accelerate_utils.py:120]: huh, these are actually just:
```
if not args.cpu and (
    args.use_fsdp or args.use_megatron_lm or args.multi_gpu
):
    multi_gpu_launcher(args)
```

File: python/ray/train/huggingface/accelerate/_accelerate_utils.py [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+            accelerate_config.setdefault("distributed_type", "NO")
+            accelerate_config.setdefault("mixed_precision", "no")
+            accelerate_config.setdefault("compute_environment", "LOCAL_MACHINE")
+            accelerate_config.setdefault("use_cpu", True)
+            config = ClusterConfig(**accelerate_config)
+            config.to_json_file(accelerate_config_path)
+        else:
+            accelerate_config_path = (
+                str(accelerate_config) if accelerate_config else default_config_file
+            )
```
Comment (gjoliver) [inline python/ray/train/huggingface/accelerate/_accelerate_utils.py:156]: wait, if we get a dict, we first write it into a file, then load it back in right after?
why don't we keep the dict, and load the file only if accelerate_config is a string path?

File: python/ray/train/huggingface/accelerate/accelerate_trainer.py [hunk 16]
```diff
@@ -0,0 +1,422 @@
...
+            # Get the rank of the worker on the current node.
+            session.get_local_rank()
+
+    For more information, see the documentation of
+    :class:`~ray.train.torch.TorchTrainer`.
+
+    .. note::
+
+        You need to use ``session.report()`` to communicate results and checkpoints
+        back to Ray Train.
```
Comment (gjoliver) [inline python/ray/train/huggingface/accelerate/accelerate_trainer.py:92]: Ray Tune?

File: python/ray/train/huggingface/accelerate/accelerate_trainer.py [hunk 17]
```diff
@@ -0,0 +1,422 @@
...
+
+            # If using GPUs, set this to True.
+            use_gpu = False
+
+            # Define NN layers archicture, epochs, and number of workers
+            input_size = 1
+            layer_size = 32
+            output_size = 1
+            num_epochs = 200
+            num_workers = 3
```
Comment (gjoliver) [inline python/ray/train/huggingface/accelerate/accelerate_trainer.py:141]: same comment here. can we not duplicate this code?

File: python/ray/train/huggingface/accelerate/accelerate_trainer.py [hunk 18]
```diff
@@ -0,0 +1,422 @@
...
+            provided datasets.
+        resume_from_checkpoint: A checkpoint to resume training from.
+    """
+
+    def __init__(
+        self,
+        train_loop_per_worker: Union[Callable[[], None], Callable[[Dict], None]],
+        *,
+        train_loop_config: Optional[Dict] = None,
+        accelerate_config: Optional[Union[dict, str, Path, os.PathLike]] = None,
```
Comment (gjoliver) [inline python/ray/train/huggingface/accelerate/accelerate_trainer.py:270]: this looks a bit weird, why is the new config param randomly placed at the 2nd place?
move it to the top?

File: python/requirements/ml/requirements_train.txt [hunk 19]
```diff
@@ -9,9 +9,17 @@ tensorboardX==2.4.1
...
+# We need transformers>=4.19.0 for HuggingFaceTrainer to work (not available on py 3.6)
+# If changing the version here, also change it in HuggingFaceTrainer docstring
 transformers==4.18.0; python_version <= '3.6'
 transformers==4.19.1; python_version > '3.6'
-accelerate==0.5.1
+# We need accelerate>=0.17.0 for AccelerateTrainer to work (not available on py 3.6)
+# If changing the version here, also change it in AccelerateTrainer docstring
+accelerate==0.5.1; python_version <= '3.6'
+accelerate==0.17.1; python_version > '3.6'
+deepspeed; python_version > '3.6'
```
Comment (gjoliver) [inline python/requirements/ml/requirements_train.txt:23]: hmm, now deepspeed will always get installed regardless of whether users actually need it or not?

File: doc/requirements-doc.txt (hunk 9 above)
Comment (Yard1) [reply doc/requirements-doc.txt:4]: This is because the python files are imported during doc build. If a library is missing, then the doc build fails. The alternative is to mock the module, but some modules like accelerate which do unorthodox stuff on import can still fail with that.

File: doc/source/ray-air/trainers.rst (hunk 12 above)
Comment (Yard1) [reply doc/source/ray-air/trainers.rst:175]: no, that's the accelerate trainer functionality

File: doc/source/ray-air/doc_code/accelerate_trainer.py (hunk 11 above)
Comment (Yard1) [reply doc/source/ray-air/doc_code/accelerate_trainer.py:55]: this is the same as in `doc/source/ray-air/doc_code/torch_trainer.py` and `doc/source/ray-air/doc_code/hvd_trainer.py`

File: python/ray/train/BUILD (hunk 13 above)
Comment (Yard1) [reply python/ray/train/BUILD:597]: the test was timing out for me, though you are correct, it should not be impacted by this. perhaps that was a fluke - I'll revert and see if it passes

File: python/ray/train/huggingface/accelerate/_accelerate_utils.py (hunk 14 above)
Comment (Yard1) [reply python/ray/train/huggingface/accelerate/_accelerate_utils.py:120]: this is the same structure as in accelerate itself (makes it easier to compare the code between the two)

File: python/ray/train/huggingface/accelerate/accelerate_trainer.py (hunk 17 above)
Comment (Yard1) [reply python/ray/train/huggingface/accelerate/accelerate_trainer.py:141]: This is the same as in the TorchTrainer docstring

File: python/ray/train/huggingface/accelerate/accelerate_trainer.py (hunk 18 above)
Comment (Yard1) [reply python/ray/train/huggingface/accelerate/accelerate_trainer.py:270]: to me it felt it was nice to have the `train_loop_per_worker` and `train_loop_config` together, like in other Trainers - those are all keyword only arguments regardless (you can't pass them positionally)

File: python/ray/train/huggingface/accelerate/_accelerate_utils.py (hunk 15 above)
Comment (Yard1) [reply python/ray/train/huggingface/accelerate/_accelerate_utils.py:156]: yes, that way we ensure that Accelerate works correctly (as it does some side effects on loading). You can't pass a dict in Accelerate itself, which is why we need to do this dance with files - otherwise we would have to use Accelerate private APIs to replicate the behavior

**Comment by Yard1** [comment]
@gjoliver in regards to doc changes, seeing as those would concern multiple docs & docstrings, how about we do that in a followup PR

File: doc/source/ray-air/doc_code/accelerate_trainer.py (hunk 10 above)
Comment (gjoliver) [reply doc/source/ray-air/doc_code/accelerate_trainer.py:19]: so they are all wrong then: https://peps.python.org/pep-0008/#constants

File: doc/source/ray-air/doc_code/accelerate_trainer.py (hunk 11 above)
Comment (gjoliver) [reply doc/source/ray-air/doc_code/accelerate_trainer.py:55]: wait, why do we want to carry "unclear" stuff into all new examples?

File: doc/source/ray-air/trainers.rst (hunk 12 above)
Comment (gjoliver) [reply doc/source/ray-air/trainers.rst:175]: so why do you start the sentence with ``This Trainer is also an extension of  :class:`TorchTrainer <ray.train.torch.TorchTrainer>, ...``, if the functionality mentioned in the 2nd half has nothing to do with it?

File: python/ray/train/huggingface/accelerate/_accelerate_utils.py (hunk 14 above)
Comment (gjoliver) [reply python/ray/train/huggingface/accelerate/_accelerate_utils.py:131]: ok

File: python/ray/train/huggingface/accelerate/_accelerate_utils.py (hunk 15 above)
Comment (gjoliver) [reply python/ray/train/huggingface/accelerate/_accelerate_utils.py:171]: I see. can you add this explanation as comments in the file? key piece of information for anyone who needs to maintain this code.

File: python/ray/train/huggingface/accelerate/accelerate_trainer.py (hunk 17 above)
Comment (gjoliver) [reply python/ray/train/huggingface/accelerate/accelerate_trainer.py:141]: come in and update all of them all at once?

**Comment by gjoliver** [comment]
> @gjoliver in regards to doc changes, seeing as those would concern multiple docs & docstrings, how about we do that in a followup PR

ah, ok, I read the comments in github order, so didn't see this until now.
yeah, followup PR sounds good to me.

**Review by gjoliver — APPROVED** [review]
ok, I think we need to do some cleanups for some of the examples.
but those can be done as a followup PR.

File: doc/source/ray-air/trainers.rst (hunk 12 above)
Comment (Yard1) [reply doc/source/ray-air/trainers.rst]: Reworded

File: python/ray/train/huggingface/accelerate/_accelerate_utils.py (hunk 15 above)
Comment (Yard1) [reply python/ray/train/huggingface/accelerate/_accelerate_utils.py:173]: Added
